From d8e2c3b72c907c3e09162da23eeffa1b09748558 Mon Sep 17 00:00:00 2001
From: lxl <lxl@xdja.com>
Date: Sat, 3 Sep 2022 14:22:38 +0800
Subject: [PATCH 09/11] fix rmnet wlan conflict

---
 system/netd/server/NetdNativeService.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/system/netd/server/NetdNativeService.cpp b/system/netd/server/NetdNativeService.cpp
index 85359f5e4e..f1eb3fb091 100755
--- a/system/netd/server/NetdNativeService.cpp
+++ b/system/netd/server/NetdNativeService.cpp
@@ -1069,8 +1069,10 @@ binder::Status NetdNativeService::networkAddRouteParcel(int32_t netId,
     const char *gw = nullptr;
     char cmd[MAX_CMD_LEN];
 
-    if (-1 != sys_id) {
+    int is_rmnet = !strncmp("rmnet", route.ifName.c_str(), strlen("rmnet"));
+    if (-1 != sys_id && is_rmnet) {
         //cells
+        system("ip rule add from all lookup main prio 31999");// NOLINT(cert-env33-c)
         //config vnet ip
         snprintf(cmd, sizeof(cmd), "ifconfig %s %s netmask %s up",
             route.ifName.c_str(),
-- 
2.17.1

